### ISSUE TITLE
Implications of non-sticky progressive arguments - should we keep it?

### ISSUE THREAD

**Issue by Jopie64** [wamp-proto/wamp-proto#450]
When looking at [progressive call arguments spec](https://wamp-proto.org/wamp_latest_ietf.html#section-11.2-25), I was surprised to read that progressive calls can also _not_ be sticky. I feel that leads to all kinds of weird implications. E.g., what happens with call cancellation?

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee 1
    participant Callee 2

    Caller->>Dealer: CALL (progress=true)
    Dealer->>Callee 1: INVOCATION
    Caller->>Dealer: CALL (still the same call instance)
    Dealer->>Callee 2: INVOCATION
    Caller->>Dealer: CANCEL
    note right of Dealer: *Confused*
```

Dealer confused because, to which callee should it route cancellation?

* _None_<br>
  Could cause leak at callees

* _Callees involved by the dealer_<br>
  Feels random, but probably the best in this scenario.

* _Callees with shared registration of this procedure_<br>
  But what if one of the callees unregisters that procedure during the call? Should dealer consider that callee done with the call already?

---
Or another scenario: when one of the callees considers the call done, should the dealer communicate that to other involved callees?

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee 1
    participant Callee 2

    Caller->>Dealer: CALL (progress=true)
    Dealer->>Callee 1: INVOCATION
    Caller->>Dealer: CALL (still the same call instance)
    Dealer->>Callee 2: INVOCATION
    Callee 1->>Dealer: YIELD(progress=true)
    Dealer->>Caller: RESULT(progress=true)
    Callee 2->>Dealer: Final YIELD
    Dealer->>Caller: Final RESULT
    note right of Dealer: Shouldn't callee 1 know this?
    Callee 1->>Dealer: YIELD(progress=true)
    note right of Dealer: Now what?
```

And if so, what should it send? Cancel? And how should it react to progressive results from other callees that are unaware of the call being done? (Which can now also happen due to races.)

---
All this isn't clear to me reading the spec...
My blunt opinion would be to remove this non-sticky progressive call option entirely, cause I also can't think of a good usecase. But if there is a good usecase, and it's worth the effort to keep this in my eyes complex feature, at least specify what happens in these kinds of scenario's...

_Originally posted by @Jopie64 in https://github.com/wamp-proto/wamp-proto/issues/448#issuecomment-1444631919_

**Comment by ecorm** [comment]
>    All this isn't clear to me reading the spec...
>    My blunt opinion would be to remove this non-sticky progressive call option entirely, cause I also can't think of a good usecase. But if there is a good usecase, and it's worth the effort to keep this in my eyes complex feature, at least specify what happens in these kinds of scenario's...

As I mentioned in the parent issue, one use case I can think of is parallel processing of real-time chunks of data.

**Comment by Jopie64** [comment]
In comment https://github.com/wamp-proto/wamp-proto/issues/448#issuecomment-1444650013, @oberstet wrote:

> yeah, we need to be precise/consistent IMO rgd terminology, e.g. one call with "progressive call arguments" and/or "progressive call results".
>
> One "progressive call" is a call with at least one of "progressive call arguments" or "progressive call results" (or both).

I used the term "progressive call" but in this case I was particularly talking about _progressive call arguments_. Especially if and how to communicate its ending when arguments are spread over multiple callees.

> And then there are calls which use "shared registrations" - that is the callee-side. There is no directly corresponding pendent "shared caller". A single call always originates from a single session.

True. Hence it's the non-sticky callee part that I want to mention :)

---

My mental model of a single call instance which is progressive on both sides (arguments and results) is like a phone call. The first `CALL` sets it up, or "dials" the other party. The dealer is the operator selecting the callee used by the address (URI) given by the caller and establishes a connection between the two parties. Now both parties can talk to each other by sending arguments and results, until one of them 'hangs up' by sending CANCEL or a final YIELD, after which the call is disconnected.

When using the non-sticky progressive arguments option, this mental model is messed up 😅. The caller makes one argument that ends up at one callee, and makes another that ends up with another callee, who has no idea of an already running call. If one of the callees 'hangs up' (final YIELD), what does that mean for the other callees? Is there order in progressive results send back to the caller? Does the caller know which callee send it?

As @ecorm said, indeed parallel processing could be a valid use case... But then still, what happens when one of the callees leaves while not done processing a chunc? How would the caller know it will never receive an answer?

Of course we could think of a way and define stuff for this in the spec. But then, for such a scenario I would not use the progressive argument feature at all, but simply make multiple individual normal non-progressive argument calls on the caller side. This has the added benefit of callees reporting progress of individual chuncs and the router is still able to route different chuncs to different callees. Also when an individual chunc fails processing, the caller would know and can decide to retry only that chunc.

So I still feel it's lacking a good use-case...

**Comment by ecorm** [comment]
> Is there order in progressive results send back to the caller? Does the caller know which callee send it?

The caller can assign its own job number to each chunk (via call arguments) that the callee repeats in the returned result. So the caller becomes responsible for correlation in this scenario when the results can be returned out of order.

> But then, for such a scenario I would not use the progressive argument feature at all, but simply make multiple individual normal non-progressive argument calls on the caller side. This has the added benefit of callees reporting progress of individual chuncs and the router is still able to route different chuncs to different callees. Also when an individual chunc fails processing, the caller would know and can decide to retry only that chunc.

Sounds reasonable to me. Like you said, the router can still be leveraged for routing chunks to/from callees, which alleviates an objection @oberstet had about clients having to perform their own routing.

> So I still feel it's lacking a good use-case...

I now agree in light of your well-put thoughts.

**Comment by ecorm** [comment]
> Jopie64: So I still feel it's lacking a good use-case...

To play the Devil's Advocate... A use case could be where parallel processing of real time data is desired, but the caller is too constrained in computing power or memory to perform manual correlation and out-of-order reassembly of the partial results.

But do we want the reassembly and fragment loss complexities to be pushed to a general-purpose WAMP router? There are entire libraries dedicated to multi-node parallelism. Is this a rabbit hole we're willing to go down in?

In this use case, a server-side WAMP "mediator" peer could be written to respond to calls from this constrained device, and forward them as regular calls to a WAMP router set up for shared registrations (say, round-robin). This "mediator" peer then takes care of reassembling the partial results before returning them back "fully cooked" to the constrained caller device.

**Comment by Jopie64** [comment]
> To play the Devil's Advocate... A use case could be where parallel processing of real time data is desired, but the caller is too constrained in computing power or memory to perform manual correlation and out-of-order reassembly of the partial results.

I don't see how non-sticky progressive call arguments feature would help here..?

Reassembly needs to be done somewhere. And if it doesn't happen on the constraint caller, why would it help then to use that feature instead of doing independent calls? In both cases results are probably forwarded to a reassembly entity. The constraint caller doesn't have to do anything extra when not using that feature, except maybe that it needs slightly more memory to maintain the individual outstanding calls. But with that little extra memory it now becomes possible to detect failure of processing a single chunk that would otherwise be impossible. (Unless we invent other features on top of that too make it even more complex.)

Also if we want to reduce even that little extra memory, callees could also be made so that they communicate results over a separate call back to the caller.

**Comment by oberstet** [comment]
> My mental model of a single call instance which is progressive on both sides (arguments and results) is like a phone call.

This is a very good mental model!

> I will go into your second part touching "shared registrations", and also how that could combine with the this "double progressive" concept, but before, let me take the chance to dive a little bit further down this "phone call" model first, explore where it leads to. I'll split up my answer into multiple comments ..

So the first note I want to make is, if you think of a single WAMP RPC that is progressive on both ends (both for call arguments and call results) as a single "phone call", this "phone call" is a *bidirectional streaming channel* (or *full duplex streaming channel*), and it is "ad hoc".

It's like a [virtual circuit](https://en.wikipedia.org/wiki/Virtual_circuit) bound to a single action (the single WAMP RPC). The circuit switching happens at layer 7 (WAMP), and the WAMP router(s) is the switching network.

BUT: you can have *multiple phone calls in parallel* to different parties! Think like: you have multiple phones that you can use in parallel.

The second note is: imagine you "Alice" are doing this "phone call" like in above with "stereo audio".

Now imagine the person "Bob" you call picks up the phone (starts replying progressive to the incoming INVOCATION), and immediately calls two other persons "Fred" and "Linda" using the same double-progressive/full-duplex RPC method.

Bob will split up the audio he receives from Alice into L-Channel and R-Channel, and replicates that on his phone calls to Fred and Alice respectively.

Bob will also mix the incoming audio from Fred and Alice into a new stereo audio feed and send that to Alice.

IOW: Bob acts as a two-way stereo audio mixer.

What is also obvious: should Fred, when he picks up the phone, himself do another outgoing call, that is fine as long as he doesn't dial into anyone in the upstream route back to Alice - if he does, the feedback will make the setup into an oscillator.

The *streaming network* that Alice, Bob, Fred and Linda span up must not have cycles. It must be a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph).

So there you have it: above is pretty much my mental model for "full duplex streaming WAMP Calls". I am thinking of them as an ad-hoc streaming network (DAG). A single step (Alice to Bob only) single party fan-in/fan-out (Alice isn't on the phone with someone else at the same time, same for Bob) is just one special case (the simplest) for a topology at that streaming network level.

**Comment by oberstet** [comment]
Ok, so here comes part 2 of my answer. There will be 3 parts in total:

1. dive into "progressive calls": the answer above pretty much dives to the bottom of this concept, explores the full potential
2. dive into "shared registrations": this answer (below)
3. dive into "shared registrations + progressive calls": that will be part 3

This last option (combination) is the tricky one - without further adding in "call canceling" already (which is just a follow up, further design drill down then)

-----

**part 2:**

While **progressive calls** expand the semantics of WAMP for single RPC calls in the *temporal* domain (one single RPC can have a sequence of call arguments, and a sequence of call results, both tied to that single RPC), **shared registrations* - without progressive calls - expand WAMP in the *spatial* domain.

A single RPC is normally routed to **one of one** registered endpoint for the procedure matching the call.

With shared registrations, a single RPC is routed to **any one of many** endpoints registered for the matching procedure.

It has 2 main use cases:

* https://wamp-proto.org/wamp_latest_ietf.html#name-load-balancing
* https://wamp-proto.org/wamp_latest_ietf.html#name-hot-stand-by

> There is more material about these 2 use cases here: https://crossbario.com/static/presentations/microservices/index.html#/41 and https://github.com/crossbario/crossbar-examples/tree/master/scaling-microservices. The latter also demos two sub-options for scale out use case available in Crossbar.io: **concurrency control** and **call queuing** which I will not go into further here.

Now, if you think about "one of one" and "one of many", the natural question to ask is: what about **all of many** and **one specific of many**?

This is hinted only in the spec here: https://wamp-proto.org/wamp_latest_ietf.html#name-sharded-registration

There is experimental code, and some spec text drafts .. forgot where exactly. But as a matter of fact: this would complete the design space in WAMP in this sub-area.

For example, I think of "All Calls" ("all of many") as basically a per single RPC *Scatter-Gather*, and "Partitioned Calls" ("one specific of many") as a form of per single RPC *Sharding*.

----

This thinking of *progressive option* as an expansion of WAMP in the *temporal domain*, and of *sharded option* as an expansion of WAMP in the *spatial domain* is useful.

Because then we can think about *progressive + sharded option* in terms of expansion of WAMP (for single RPCs) in the *spatio-temporal domain*, and what that actually means. And then we should be able to see how "canceling" should behave.

I'll come back to this in part 3 ... later the day;) Sorry for the massive text, but I think we now have actually reached "the core" =)

![grafik](https://user-images.githubusercontent.com/233340/221353100-92535f82-d597-43e3-b1c8-42cedf30a7ed.png)

**Comment by oberstet** [comment]
now coming back to **part 3**, that is "progressive call arguments + shared registrations".

I now actually think it should work like: _first_ spatial distribution (from the shared registration), and once *one of many* endpoint has been selected by the Dealer, _then_ the complete sequence of progressive call arguments (the temporal dimension) MUST go to the same selected *one* endpoint

> I was surprised to read that progressive calls can also not be sticky

**I agree!**

It should be: first spatial distribution, then temporal spreading. It doesn't make sense to combine both.

That means, it will be "sticky" always, and the [Progressive Calls and Shared Registration](https://wamp-proto.org/wamp_latest_ietf.html#section-11.2-25) section then would need to be adjusted to explain above.

> E.g., what happens with call cancellation?

This question disappears with above. The dealer will need to keep track of the single RPC and the selected *one of many* endpoint, to forward subsequent progressive call arguments as well as a CANCEL

----

I would further argue that the current name "Progressive Calls"

https://wamp-proto.org/wamp_latest_ietf.html#name-progressive-calls

for what is more precisely "Progressive Call Arguments" is confusing.

In that more precise terminology, "Progressive Calls" are calls which are either using "Progressive Call Arguments" or "Progressive Call Results" or both.

----

The most important learning for me actually comes from: why didn't we notice earlier, why does it come up now?

From my side, the reason is: while Crossbar.io implements all of the "shared registrations" stuff (even including "max concurrency" and "call queuing", and it implements "progressive call results", it currently does *not* implement "progressive call arguments"!!

That means, I didn't had working code to test the design against.

I _think_ (not sure, forgot): while @ecorm 's router implements _both_ "progressive call results" _and_ "progressive call arguments", it does _not_ implement "shared registrations".

That means: neither of us had working code for the full-blown _combination_!

I think, even though I was slightly aware of that, I nevertheless insisted to add this sticky or not thing to "progressive call arguments"?

In summary (note to self): never come up with a design or design aspects without working code!

**Comment by Jopie64** [comment]
Eagerly awaiting part 3 😊
But already a short reaction to part 2

> This last option (combination) is the tricky one - without further adding in "call canceling" or "sticky" already (which are then just follow ups, further design drill downs)

I think there isn't any trickiness when progressive argument calls are sticky, even with shared registrations. The callee is selected in the first call, and progressive arguments are sent to the same callee.
* It's clear where progressive arguments go to
* It's clear where (progressive) responses come from
* It's clear what happens when that callee leaves
* It's clear what happens when other (not selected) callees leave (nothing)
* It's clear what needs to be done when the caller cancels the call
* It's clear the call just ends when the callee sends a final yield or error

The trickiness arize when progressive arguments are _not_ sticky.

I think similar implications happen with those sharded calls, progressive arguments or not.
I feel the combination of the two features won't add more complexity and can be handled independently, but let's see...

**Comment by Jopie64** [comment]
Ah I was too soon. You already posted part three with the same conclusion 😅

Anyway, as said, probably similar trickiness will arise with sharded registrations, but that will be a different subject/issue...

So about the non sticky progressive call arguments, conclusion is that we wait for consensus to drop this option? Or any other ideas?

**Comment by ecorm** [comment]
> for what is more precisely "Progressive Call Arguments" is confusing.
>
> In that more precise terminology, "Progressive Calls" are calls which are either using "Progressive Call Arguments" or "Progressive Call Results" or both.

How about this?
- Progressive Calls -> Caller-to-Callee Streams
- Progressive Call Results -> Callee-to-Caller Streams
- Both -> Bidirectional Streams

In my client implementation, I plan on providing a streaming API where the user need not even be aware that it uses `CALL` messages under the hood.

> I think (not sure, forgot): while ecorm 's router implements both "progressive call results" and "progressive call arguments", it does not implement "shared registrations".

That is correct. "Implements" is too kind a word because barely anything is tested yet on the router side. I don't know whether I'll ever implement shared registrations, unless there's a 3rd party demand for it and I implement it in my spare time.

**Comment by oberstet** [comment]
> How about this?

-1: this is even more confusing. those are remote procedure calls with specific argument/result types, that is stream typed arguments and/or stream typed results. the "stream type" itself has no external identity, while the "procedure" has (e.g. authorization)

thus, the naming "RPC with progressive call arguments" and/or "RPC with progressive call results" or "progressive RPCs" seems better. in my eyes. 

if we don't change "progressive calls" - which are currently only calls with progressive call arguments - then how do we name calls which use _both_ progressive call arguments _and_ progressive call results?

as far as from my PoV, we need to either change the naming, or answer the 2nd question ...

> That is correct.

ok, right. thanks for confirming!

**Comment by KSDaemon** [comment]
How about this variant?

* `Progressive call results` - seem to be clear.
* `Progressive calls` (as they are described right now) → rename to `Progressive invocations`? We already have this term and it means the caller-to-callee direction.
* Use `Progressive calls` term for kind of generalized solution when one or both flows are progressive.

**Comment by ecorm** [comment]
> those are remote procedure calls with specific argument/result types, that is stream typed arguments and/or stream typed results.

I humbly disagree that they are remote procedure calls. They don't behave at all like their programming language counterparts. They just happen to use `CALL` and `RESULT` messages under the hood, but it could have just as easily been designed to use similar messages dedicated to streaming.

You were fond of Jopie64's analogy to phone calls, but those model streams and not RPCs. A real-world analogy to RPCs would be letters sent by the post where a reply (or delivery failure notice) is expected.

Regarding "typed arguments" in streams (in general), where is it defined what format (or lack thereof) chunks must have? To me, the concept of a stream is that data is transferred in chunks. It is up to the protocol/application to define what the format is for these chunks. Some may have loose formats and others may have rigid formats.

I do admit though that it's strange for a stream chunk to have keyword arguments. This is a side-effect of choosing `CALL` and `RESULT` for the underlying messages, where the payloads have function argument semantics.

> the "stream type" itself has no external identity, while the "procedure" has (e.g. authorization)

The stream identities (I've seen it called "channel" in other places) in WAMP are URIs and request IDs, just like in regular RPCs. Streams are ephemeral (just like procedure call requests) and their identities happen to share the same namespace as "normal" RPCs. Streams are established by the initial `CALL` message, and are terminated with either the final `CALL` or the final `RESULT`. Stream listeners (callees) use the procedure URI to register their interest in establishing a stream terminus. Shared registrations would be a way of multiplexing the stream to multiple termini.

I now see progressive call arguments and progressive call results for what they really are: streams. I'm going to change CppWAMP's API to reflect this reality and I'm confident I'm going to get a friendlier API as a result.

I understand that the terms "progressive call arguments/results" have been in use for a long time in the WAMP community, and I don't expect there to be much willingness to make drastic changes to these terms (this also happens in code where if you wait too long, collaborators will resist name changes even if the new name is better than the old one). Anyway, I'm happy to keep using the existing (or derivative) terminology that the WAMP community has already been accustomed to.

Sorry for the wall of text, but this has been a way for me to think out loud about adopting a stream API in my client that hides the complexities of the progressive call arguments/results feature.

**Comment by oberstet** [comment]
> You were fond of Jopie64's analogy to phone calls, but those model streams and not RPCs.

No, that's exactly the point: to have a phone call, a telephone network first sets up a virtual circuit. This is done via signaling. Once the circuit is setup, then there is a payload feed that transports media.

Transferred to WAMP: to have a phone call, you first need to do a RPC with progressive call arguments and progressive call results enabled. Once that is done (the first CALL/INVOCATION has hit the other peer), each of the parties use progressive arguments/results for media.

> where is it defined what format (or lack thereof) chunks must have?

that's the beauty: we already have implicit framing! because one call at the app level will result in one CALL (progress=true) or YIELD (progress=true)

> The stream identities (I've seen it called "channel" in other places) in WAMP are URIs and request IDs

No. The signaling phase in the phone call analogy is mapped/piggyback to the first CALL and first YIELD

----

let me try another angle to convince you of the value of making a distinction between "outer RPC" (signaling) and "inner RPC (media stream)"

take Rust. there in an explicit type (a trait) for streams:

https://docs.rs/futures/latest/futures/stream/trait.Stream.html

but how do you get a duplex pair of streams?

you need signaling - out-of-band or handshake-prefixed in-band on the wire, and you need to expose the signaling to the app using some "remote function call" thingy ...

anyways. at least this is my mental model. works for me, I won't change it.

**Comment by ecorm** [comment]
@oberstet Heh, you've just convinced me even _more_ that the progressive call arguments/results feature are just streams in disguise. :grin:

In any case, they _allow_ me to implement streams, which I think was the motivation for those features in the first place, right?

The C++ RAII feature is going be useful for implementing the stream termini, where if they go out of scope, they automatically terminate the stream cleanly via final `CALL` or `RESULT` message. I'm actually looking forward to implement this idea; it looks like fun.

**Comment by Jopie64** [comment]
I see it like this:
You have different kinds of streams, like media and event streams. You also have different kinds of transports (on different layers) like TCP, websocket, wamp, http, sip and RTP. Some of them are suitable for streams, some more for event streams or signalling (e.g. sip) and some more for media streams (e.g. RTP).
So in this view, WAMP progressive calls is defining a (higher layer) transport suitable for full duplex event streams.
So indeed you could implement c++ streams using WAMP progressive calls (although c++ streams are more like media streams). But you could also implement observable (Rx), C# IAsyncIterable or js callbags over them.

Stream concept is just something you can implement over certain transports, but it's not a transport in itself.

Note:
> they automatically terminate the stream cleanly via final CALL or RESULT message

Final RESULT will terminate the call entirely but final CALL will not, it will just end the caller sending part (or mute the caller so to speak). To terminate the entire call from the caller side you have to send a CANCEL.

**Comment by Jopie64** [comment]
> Stream concept is just something you can implement over certain transports, but it's not a transport in itself.

Now I think of it, you can _use_ full duplex streams as transports if you want. So you can implement WAMP over websocket, but in theory also over a WAMP progressive call. So you can have a WAMP callee act as a higher level WAMP router. And on top of that you can do it again. 🤯 Inception.

**Comment by oberstet** [comment]
> And on top of that you can do it again.

yes, that (tunneling) is generally always possible, like "TCP-over-WAMP-Progressive-Calls" or "WAMP-over-WAMP-Progressive-Calls" - but not always useful;)

thinking about: "WAMP-over-WAMP-Progressive-Calls" => this is actually running embedded full WAMP session in ad-hoc Progressive-Calls at the base layer

personally, I find Wireguard very interesting. it totally destroys IPSec and OpenVPN and all the proprietory shenanigans

Wireguard is tunneling IP over UDP, which for various reasons is really powerful (these days). it also, on Linux, perfectly integrates via user-level network interfaces (TUN) which allows to reuse the TCP network stack already in Linux without duplication

**Comment by oberstet** [comment]
> that the progressive call arguments/results feature are just streams in disguise.

you will be shocked: in the end, it's all bits and bytes!

or, as Kronecker said: "Die ganzen Zahlen hat der liebe Gott gemacht, alles andere ist Menschenwerk."

I admit, I have a bit of an inclination for constructivism;)

**Comment by oberstet** [comment]
> they automatically terminate the stream cleanly via final CALL or RESULT message

there are two streams in a progressive call:

* progressive call arguments
* progressive call results

each of the streams can be "closed" at two ends (Peers) with

* CALL progress == false
* RESULT progress == false

at the *Caller* end, or with

* INVOCATION progress == false
* YIELD progress == false

at the *Callee* end

a progressive call is closed only when both streams are closed

**Comment by Jopie64** [comment]
> a progressive call is closed only when both streams are closed

Just for my understanding:
So does that mean a progressive arguments call is _not_ always closed when it receives the (final) YIELD/RESULT? Only when it _also_ received the final argument?

Makes sense now I think about it...

But still both the caller and the callee can close both ends at the same time by sending a CANCEL or ERROR respectively right?

**Comment by oberstet** [comment]
> Just for my understanding: So does that mean a progressive arguments call is not always closed when it receives the (final) YIELD/RESULT? Only when it also received the final argument?

you are right, we need to discuss that also! tbh, I've overlooked this aspect.

I would start argue what we aim for ... make a fully "symmetric" two-way streaming channel ride on top of a single RPC

Now, I you consider the callee/router side sending a final YIELD/RESULT, this will make the router:

* no accept any further YIELDs from the callee side, and the router never send any further message to the caller with the same request ID
* the router will consider the call "completed", so it wouldn't forward any more CALL arguments it receives from the caller
* the router will need to remember the request ID for the progressive call so that race conditions, anything further arriving from the caller, can be dealt with (ignored)

now, how should the caller side work?

I would argue:

* without progressive call arguments (or "progress==false" on the first and only CALL message), the router will keep the call open until it received the last (or only) YIELD with "progress==false"
* IOW: the call is completed from the callee side

mmh. but if so, how can a caller then actively end / complete a call (when "progress==false" isn't that signal)?

maybe via CANCEL?

**Comment by Jopie64** [comment]
> thinking about (not sure though):
>
> * Bob hangs up by sending a last YIELD with progress==false
> * Alice hangs up by sending a last CALL with progress== false and a final CANCEL

That was my first assumption indeed.
But then, what's the use of the final CALL?
It could just simply end the call with a CANCEL...

I think there are 2 options.

1. This above: The call ends with a final YIELD/RESULT.

  If we do that, I'd argue that once announced with REGISTER (and maybe first CALL, not sure even that is needed), the call is always a progressive arguments call. When the caller wants to send a single argument, it just does so, but if it wants to, it can send more. There's no need to communicate in a standardized way it doesn't send anymore arguments since it doesn't have any WAMP-level effect anyway.

2. The call ends when a final YIELD/RESULT _and_ a final argument is sent

So when the final YIELD is sent the call is not done yet. Only after the final argument is also sent.
In that case the progress=true/false of subsequent arguments do carry WAMP-level value and it should stay.

---

Thinking of a use case for the second:
Caller sends ephemeral location information to the callee for which no confirmation is needed. Callee just YIELDs OK with final YIELD and starts grokking location data.
When the caller is done, it just ends with the final CALL and the callee doesn't have to react anymore.

Note:
I think in both cases, the call is always ended with CANCEL from caller or ERROR from callee. No matter whether final things were already sent from one end or not.

**Comment by oberstet** [comment]
> In any case, they _allow_ me to implement streams, which I think was the motivation for those features in the first place, right?

yes:

* make an adhoc fully "symmetric" two-way streaming channel ride on top of a single RPC
* use the RPC call both as signaling and for media transport
* allow for multiple concurrent such channels (both from and to single sessions)

> But then, what's the use of the final CALL?

Good behavior? Both peers can drop the transport, eg TCP connection, anytime anyways. Both can not drop the TCP but just refuse to or be unable to reply or send anything.

**Comment by ecorm** [comment]
> But then, what's the use of the final CALL?

Consider the case where the Callee performs a computation using progressive CALL arguments, including the final one. The Caller then waits for the final RESULT of the computation because it needs that final result.

I apologize for derailing the issue's topic of non-sticky progressive arguments with my streaming idea, but some very good discussion came out of it as a result.

**Comment by KSDaemon** [comment]
> the call is always a progressive arguments call. When the caller wants to send a single argument, it just does so, but if it wants to, it can send more. There's no need to communicate in a standardized way it doesn't send anymore arguments since it doesn't have any WAMP-level effect anyway.

I don't like this idea. In this case, Callee can not say for sure did it receive all the data or only a chunk. Marking call explicitly is good because Callee definitely knows the current state.

**Comment by KSDaemon** [comment]
Regarding the main topic: okay, this is my bad. I thought about how to bind the call to a specific peer in the shared RPC environment. But I didn't finalize that it doesn't make sense to do otherwise and it carries a lot of problems and complications. Thank you all, we figured it out and can seem to fix it easily. Even my implementation in Nexus now only supports the sticky version. so if we just take that option out, nothing critical will happen. 
Why don't we do that?

**Comment by Jopie64** [comment]
@KSDaemon 

> In this case, Callee can not say for sure did it receive all the data or only a chunk. Marking call explicitly is good because Callee definitely knows the current state.

If it's important to know for the callee, (which is not always the case), it can be communicated on application level (via the argument payload).
Since on WAMP level it doesn't have any effect (when we go for option 1)  we don't need to force that info to be standardized and available in all cases, even when it's not needed.

**Comment by KSDaemon** [comment]
> If it's important to know for the callee, (which is not always the case), it can be communicated on application level (via the argument payload).

Well. Most of the features may be implemented on the application level :) But having them on the protocol and thus implemented by someone else (wamp implementation author) is cool from many perspectives.

Another view on this: right now we have a bunch of client and router implementations. In the current state when this feature (progressive calls) must be announced and clients explicitly have to set some options to use it - it is easy to check for this feature support on any side. If we go the way without the need for these options - how peer that does not support it can distinguish regular calls and progressive ones? Or even this: how the router that knows that callee is not supporting progressive calls (not announced via handshake) can prevent sending a progressive call from a Caller? So instead of processing this on the protocol level, we move this to the application level.

**Comment by Jopie64** [comment]
>> But then, what's the use of the final CALL?
>
> Good behavior? Both peers can drop the transport, eg TCP connection, anytime anyways. Both can not drop the TCP but just refuse to or be unable to reply or send anything.

Not sure if I understand correctly... Do you mean, leaving out the final CALL (when using  option 1) is comparable with TCP not sending a close signal but just don't do anything anymore when they want to end the stream?
In that case, I don't think that is comparable because the close signal does have an effect on the TCP protocol itself, not only on the application. It can clean up buffers and free port allocations etc.

When we go for option 2, the last CALL concept has meaning on WAMP level, so then it must be there. But when we go for option 1, the last CALL concept has no meaning on WAMP level. So then I don't see a reason to force it to be specified anyway...

**Comment by Jopie64** [comment]
@KSDaemon 
> Well. Most of the features may be implemented on the application level :) But having them on the protocol and thus implemented by someone else (wamp implementation author) is cool from many perspectives.

That may be a good reason. Reducing application level re-implementations. But I doubt this "last progressive call feature" is used so much that it's worth specifying at WAMP level when WAMP itself doesn't even need it.

E.g. when you use progressive call arguments to regulate back pressure or flow control, then "last progressive call" doesn't even have any meaning on application level.

> In the current state when this feature (progressive calls) must be announced and clients explicitly have to set some options to use it - it is easy to check for this feature support on any side. If we go the way without the need for these options - how peer that does not support it can distinguish regular calls and progressive ones?

I'm _not_ arguing to remove progressive call feature announcement. That is useful indeed also for backward compatibility.

I'm just arguing to remove the need for stating that a CALL is final or not, if it bears no meaning on WAMP level anyway.

**Comment by Jopie64** [comment]
Out of topic comment

@oberstet 
> personally, I find Wireguard very interesting. it totally destroys IPSec and OpenVPN and all the proprietory shenanigans

I just took a look at it and it looks very promising indeed!
Secure, fast and especially so simple!

It's often very hard to make things simple, but when that's done it's so pleasing.

**Comment by ecorm** [comment]
> I'm just arguing to remove the need for stating that a CALL is final or not, if it bears no meaning on WAMP level anyway.

I understand what you mean now. Other than CANCEL, there is currently no way for the Caller to say "shut up, I'm done" to the Callee that can be enforced by the Dealer.

The only thing a Dealer can possibly do after receiving a final CALL is disallow the Caller to send another progressive CALL. Until the Callee replies with the final RESULT, I can't think of any resources the Dealer can clean up upon receiving the final CALL.

The Caller can't do any clean-up either until it receives the final RESULT or an ERROR triggered by cancellation.

However, on the Callee side, the client library may interpret an INVOCATION as being final and provide a means for the handler to retrieve this information (or dispatch a different/separate handler). The Callee client library would not be aware any application-specific way to indicate final CALLs.

**Comment by ecorm** [comment]
> Me: However, on the Callee side, the client library may interpret an INVOCATION as being final and provide a means for the user handler to retrieve this information (or dispatch a different/separate user handler). The Callee client library would not be aware any application-specific way to indicate final CALLs.

This would pair up well with the Caller client library providing the same convenience to the user for indicating "end-of-stream".

Caller pseudocode:
```c++
auto channel = session.startStream("my_uri", options); // Sends signalling CALL with progress=true
channel.send(chunk1); // Sends CALL with progress=true
channel.send(chunk2); // Sends CALL with progress=true
channel.sendFinal(chunk3);  // Sends CALL with progress=false
```

Callee pseudocode:
```c++
session.registerStreamHandler(
    "my_uri",
    options,
    [](Payload payload, bool isFinal) {/*process payload*/});
```

Without the `Call.Options.progress` option, the client library would have to intrusively inject finality into the user's chunk payloads.

So please leave the final CALL feature in, I plan to use it myself in the upcoming streaming API of my WAMP client library!

**Comment by Jopie64** [comment]
@ecorm 
Ok I see it in a different light now, thanks! I'll ponder on that for a bit. But this indeed might ply to keep it in 😊

**Comment by Jopie64** [comment]
Should we close this issue? Non-sticky option has been removed from the docs right?

Maybe the only thing left is which behavior to choose in relation to ending the progressive call as described here: https://github.com/wamp-proto/wamp-proto/issues/450#issuecomment-1445401716

I feel like we should go for option 1 with possibility to end the progressive call invocation part for which @ecorm gave a good usecase I think. If it needs more discussion I'd suggest to do that in a new issue.

**Comment by KSDaemon** [comment]
Agree. As the essence of the issue is resolved let's close it.

**Comment by ecorm** [comment]
@Jopie64 , @oberstet , @KSDaemon , please check out #463 where I elaborate on the concept of streaming.

**Comment by ecorm** [comment]
@Jopie64 , @KSDaemon , obserstet has kindly enabled the Discussion feature of this repo, so my streaming idea post can now be found in #465 .
